backend/app/user_handler.go

```diff
@@ -0,0 +1,325 @@
+package app
+
+import (
+	"kubecloud/internal"
+	"kubecloud/models"
+	"net/http"
+	"time"
+
+	"github.com/gin-gonic/gin"
+	"github.com/rs/zerolog/log"
+	"gorm.io/gorm"
+)
+
+// Handler struct holds configs for all handlers
+type Handler struct {
+	tokenManager internal.TokenManager
+	db           models.DB
+	config       internal.Configuration
+}
+
+// NewHandler create new handler
+func NewHandler(tokenManager internal.TokenManager, db models.DB, config internal.Configuration) *Handler {
+	return &Handler{
+		tokenManager: tokenManager,
+		db:           db,
+		config:       config,
+	}
+}
+
+// RegisterInput struct for data needed when user register
+type RegisterInput struct {
+	Name            string `json:"name" binding:"required" validate:"min=3,max=20"`
+	Email           string `json:"email" binding:"required" validate:"mail"`
+	Password        string `json:"password" binding:"required" validate:"password"`
+	ConfirmPassword string `json:"confirm_password" binding:"required" validate:"password"`
+}
+
+// LoginInput struct for login handler
+type LoginInput struct {
+	Email    string `json:"email" binding:"required" validate:"mail"`
+	Password string `json:"password" binding:"required" validate:"password"`
+}
+
+// RefreshTokenInput struct when user refresh token
+type RefreshTokenInput struct {
+	RefreshToken string `json:"refresh_token" binding:"required"`
+}
+
+// EmailInput struct for user when forgetting password
+type EmailInput struct {
+    Email string `json:"email" validate:"required,email"`
+}
+
+
+// VerifyCodeInput struct takes verification code from user
+type VerifyCodeInput struct {
+	Email string `json:"email" binding:"required"`
+	Code  int    `json:"code" binding:"required"`
+}
+
+// ChangePasswordInput struct for user to change password
+type ChangePasswordInput struct {
+	Email           string `json:"email" binding:"required"`
+	Password        string `json:"password" binding:"required" validate:"password"`
+	ConfirmPassword string `json:"confirm_password" binding:"required" validate:"password"`
+}
+
+// RegisterHandler registers user to the system
+func (h *Handler) RegisterHandler(c *gin.Context) {
+	var request RegisterInput
+
+	// check on request format
+	if err := c.ShouldBindJSON(&request); err != nil {
+		log.Error().Err(err).Send()
+		c.JSON(http.StatusBadRequest, gin.H{"error": "Invalid request format"})
+		return
+	}
+
+	// password and confirm password should match
+	if request.Password != request.ConfirmPassword {
+		c.JSON(http.StatusBadRequest, gin.H{"error": "password and confirm password don't match"})
+		return
+	}
+
+	// check if user previously exists
+	_, err := h.db.GetUserByEmail(request.Email)
```

Comment:
if you added a verify mail step then you will need to check if user exists but not verified